File: 1547/main.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println(minCost(9, []int{5,6,1,4,2}))
}

func minCost(n int, cuts []int) int {
	m := len(cuts)
	cuts = append([]int{0}, cuts...)
	cuts = append(cuts, n)
	sort.Ints(cuts)
	f := make([][]int, m+2)
	for i := range f {
		f[i] = make([]int, m+2)
	}
	for i := m; i >= 1; i-- {
		for j := i; j <= m; j++ {
			if i == j {
				f[i][j] = 0
			} else {
				f[i][j] = int(^uint(0) >> 1)
			}
			for k := i; k <= j; k++ {
				f[i][j] = min(f[i][j], f[i][k-1]+f[k+1][j])
			}
			f[i][j] += cuts[j+1] - cuts[i-1]
		}
	}
	return f[1][m]
}

File: 1547/main.rs
struct Solution {}

impl Solution {
    pub fn min_cost(n: i32, cuts: Vec<i32>) -> i32 {
        let mut cuts = cuts;
        cuts.push(0);
        cuts.push(n);
        cuts.sort();
        let m = cuts.len();
        let mut f = vec![vec![0; m]; m];
        for i in (0..m).rev() {
            for j in i+2..m {
                f[i][j] = i32::MAX;
                for k in i+1..j {
                    f[i][j] = std::cmp::min(f[i][j], f[i][k] + f[k][j]);
                }
                f[i][j] += cuts[j] - cuts[i];
            }
        }
        f[0][m-1]
    }
}

fn main() {
    println!("{}", Solution::min_cost(9, vec![5,6,1,4,2]));
}